File: Subsets.py
class Solution:

    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        from copy import deepcopy
        result = []

        def getResult(curr, end, tmpResult):
            next_loc = curr + 1
            if curr == end:
                result.append(deepcopy(tmpResult))
            else:
                getResult(next_loc, end, tmpResult)
                tmp = deepcopy(tmpResult)
                tmp.append(nums[curr])
                getResult(next_loc, end, tmp)
        getResult(0, len(nums), [])
        return result

    def subsets2(self, nums):
        result = [[]]
        for num in nums:
            result.extend([item + [num] for item in result])
        return result


if __name__ == '__main__':
    a = Solution()
    print(a.subsets3([1, 2, 3]))

File: Subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
  let result = [
    []
  ];
  nums.forEach((num) => {
    result.push(...result.map(p => p.concat(num)))
  })
  return result;
};

// 思路和python列表推导相同，将每个数组的元素进行遍历，然后将遍历的元素添加到item中去